### PR TITLE
SONARJS-601 Simplify the API to create new issues

### PR DESCRIPTION
**Pull request by vilchik-elena**

### CODE REVIEW

File: sonar-compat/src/main/java/org/sonar/javascript/issues/PreciseIssueCompat.java [hunk 1]
```diff
@@ -19,61 +19,43 @@
...
 import org.sonar.api.batch.fs.InputFile;
 import org.sonar.api.batch.fs.TextRange;
 import org.sonar.api.batch.sensor.issue.NewIssue;
 import org.sonar.api.batch.sensor.issue.NewIssueLocation;
 import org.sonar.api.rule.RuleKey;
+import org.sonar.plugins.javascript.api.visitors.PreciseIssue;
 import org.sonar.plugins.javascript.api.visitors.IssueLocation;
 
-public class PreciseIssue {
+public class PreciseIssueCompat {
```
Comment (SonarTech) [inline sonar-compat/src/main/java/org/sonar/javascript/issues/PreciseIssueCompat.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1118)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/JavaScriptCheck.java [hunk 2]
```diff
@@ -30,5 +31,12 @@
 
   TreeVisitorContext getContext();
 
+  /**
+   *  This method should be deprecated, as soon as this plugin will be migrated on 5.X LTS (with support of precise issue locations)
+   *  Instead please use <code>getContext().addIssue()</code>
+   */
+  @Deprecated
+  void addLineIssue(JavaScriptCheck check, Tree tree, String message);
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/JavaScriptCheck.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1133)
MAJOR Add the missing @deprecated Javadoc tag. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* MAJOR 2 major
* MINOR 1 minor
* INFO 1 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Add a new line at the end of this file.](https://github.com/SonarSource/sonar-javascript/blob/e00218d14fa6d2679d21e3ea2c019e6d12b233bf/javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/visitors/Issue.java) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS00113)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tests/TreeCheckTest.java [hunk 3]
```diff
@@ -19,25 +19,90 @@
...
+    } catch (RecognitionException e) {
+      if ("ParsingErrorCheck".equals(check.getClass().getSimpleName())) {
+        issues .add(new LineIssue(null, e.getLine(), e.getMessage()));
+      }
+    }
+
+    return getCheckMessages(issues);
+  }
+
+  private Collection<CheckMessage> getCheckMessages(List<Issue> issues) {
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tests/TreeCheckTest.java]: MAJOR Make "getCheckMessages" a "static" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tests/TreeCheckTest.java [hunk 4]
```diff
@@ -19,25 +19,90 @@
...
+        scriptTree,
+        file,
+        symbolModel,
+        settings(),
+        new ComplexityVisitor()
+      );
+
+      issues = check.scanFile(context);
+
+    } catch (RecognitionException e) {
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tests/TreeCheckTest.java]: CRITICAL Either log or rethrow this exception. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1166)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tests/TreeCheckTest.java [hunk 5]
```diff
@@ -19,25 +19,90 @@
...
+      if ("ParsingErrorCheck".equals(check.getClass().getSimpleName())) {
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tests/TreeCheckTest.java]: CRITICAL Use an "instanceof" comparison instead. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1872)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* CRITICAL 2 critical
* MAJOR 2 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'javascript' package (org.sonarsource.javascript:javascript-frontend:src/main/java/org/sonar/javascript) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1228)

File: javascript-checks/src/main/java/org/sonar/javascript/checks/tests/TestUtils.java [hunk 6]
```diff
@@ -17,30 +17,39 @@
...
+import org.sonar.javascript.tree.symbols.SymbolModelImpl;
 import org.sonar.javascript.tree.symbols.type.JQuery;
-import org.sonar.plugins.javascript.api.JavaScriptCheck;
-import org.sonar.squidbridge.api.CheckMessage;
+import org.sonar.plugins.javascript.api.symbols.SymbolModel;
+import org.sonar.plugins.javascript.api.tree.ScriptTree;
+import org.sonar.plugins.javascript.api.tree.Tree;
 
-public class TreeCheckTest {
+public class TestUtils {
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/tests/TestUtils.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1118)

File: javascript-checks/src/main/java/org/sonar/javascript/checks/tests/TreeCheckTest.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    } catch (RecognitionException e) {
+      if ("ParsingErrorCheck".equals(check.getClass().getSimpleName())) {
+        issues .add(new LineIssue(null, e.getLine(), e.getMessage()));
+      }
+    }
+
+    return getCheckMessages(issues);
+  }
+
+  private Collection<CheckMessage> getCheckMessages(List<Issue> issues) {
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/tests/TreeCheckTest.java]: MAJOR Make "getCheckMessages" a "static" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

File: javascript-checks/src/main/java/org/sonar/javascript/checks/tests/TreeCheckTest.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+  public Collection<CheckMessage> getIssues(String relativePath, JavaScriptCheck check) {
+    File file = new File(relativePath);
+    List<Issue> issues = new ArrayList<>();
+
+    try {
+      JavaScriptCheckContext context = TestUtils.createContext(file);
+      issues = check.scanFile(context);
+
+    } catch (RecognitionException e) {
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/tests/TreeCheckTest.java]: CRITICAL Either log or rethrow this exception. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1166)

File: javascript-checks/src/main/java/org/sonar/javascript/checks/tests/TreeCheckTest.java [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+      if ("ParsingErrorCheck".equals(check.getClass().getSimpleName())) {
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/tests/TreeCheckTest.java]: CRITICAL Use an "instanceof" comparison instead. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1872)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues:
* CRITICAL 2 critical
* MAJOR 4 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'tests' package (org.sonarsource.javascript:javascript-checks:src/main/java/org/sonar/javascript/checks/tests) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'javascript' package (org.sonarsource.javascript:javascript-frontend:src/main/java/org/sonar/javascript) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1228)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* CRITICAL 1 critical
* MAJOR 2 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'tests' package (org.sonarsource.javascript:javascript-checks:src/main/java/org/sonar/javascript/checks/tests) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'javascript' package (org.sonarsource.javascript:javascript-frontend:src/main/java/org/sonar/javascript) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1228)

**Comment by pynicolas** [comment]
This is an important change in the API. I created a ticket to track that change: https://jira.sonarsource.com/browse/SONARJS-601. We will need to communicate about it when opening a vote for a new release.

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* CRITICAL 1 critical
* MAJOR 2 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'tests' package (org.sonarsource.javascript:javascript-checks:src/main/java/org/sonar/javascript/checks/tests) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'javascript' package (org.sonarsource.javascript:javascript-frontend:src/main/java/org/sonar/javascript) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1228)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* CRITICAL 1 critical
* MAJOR 2 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'tests' package (org.sonarsource.javascript:javascript-checks:src/main/java/org/sonar/javascript/checks/tests) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'javascript' package (org.sonarsource.javascript:javascript-frontend:src/main/java/org/sonar/javascript) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1228)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* CRITICAL 1 critical
* MAJOR 2 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'tests' package (org.sonarsource.javascript:javascript-checks:src/main/java/org/sonar/javascript/checks/tests) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'javascript' package (org.sonarsource.javascript:javascript-frontend:src/main/java/org/sonar/javascript) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1228)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* CRITICAL 1 critical
* MAJOR 2 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'tests' package (org.sonarsource.javascript:javascript-checks:src/main/java/org/sonar/javascript/checks/tests) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'javascript' package (org.sonarsource.javascript:javascript-frontend:src/main/java/org/sonar/javascript) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1228)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* CRITICAL 1 critical
* MAJOR 2 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'tests' package (org.sonarsource.javascript:javascript-checks:src/main/java/org/sonar/javascript/checks/tests) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'javascript' package (org.sonarsource.javascript:javascript-frontend:src/main/java/org/sonar/javascript) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1228)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* CRITICAL 1 critical
* MAJOR 2 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'tests' package (org.sonarsource.javascript:javascript-checks:src/main/java/org/sonar/javascript/checks/tests) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'javascript' package (org.sonarsource.javascript:javascript-frontend:src/main/java/org/sonar/javascript) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1228)

**Comment by pynicolas** [comment]
@vilchik-elena For JavascriptCheck, I think I would prefer to replace `addLineIssue` with `newLineIssue` which would return a `LineIssue` the same way `newIssue` returns a `PreciseIssue`. I'm not sure that creating issues on lines should be deprecated, it may make sense to create issues at line level for some rare cases.
What about issues at file level?

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* CRITICAL 1 critical
* MAJOR 2 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'tests' package (org.sonarsource.javascript:javascript-checks:src/main/java/org/sonar/javascript/checks/tests) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'javascript' package (org.sonarsource.javascript:javascript-frontend:src/main/java/org/sonar/javascript) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1228)

**Pull request merged**